rpcs3/Emu/Cell/Modules/cellCrossController.cpp

```diff
@@ -1,38 +1,76 @@
-#include "stdafx.h"
+﻿#include "stdafx.h"
 #include "Emu/System.h"
 #include "Emu/Cell/PPUModule.h"
 #include "cellSysutil.h"
-
+#include "cellCrossController.h"
 
 
 LOG_CHANNEL(cellCrossController);
 
-enum
+template <>
+void fmt_class_string<CellCrossControllerError>::format(std::string& out, u64 arg)
 {
-	CELL_CROSS_CONTROLLER_STATUS_INITIALIZED = 1,
-	CELL_CROSS_CONTROLLER_STATUS_FINALIZED = 2
-};
+	format_enum(out, arg, [](CellCrossControllerError value)
+	{
+		switch (value)
+		{
+		STR_CASE(CELL_CROSS_CONTROLLER_ERROR_CANCEL);
+		STR_CASE(CELL_CROSS_CONTROLLER_ERROR_NETWORK);
+		STR_CASE(CELL_CROSS_CONTROLLER_ERROR_OUT_OF_MEMORY);
+		STR_CASE(CELL_CROSS_CONTROLLER_ERROR_FATAL);
+		STR_CASE(CELL_CROSS_CONTROLLER_ERROR_INVALID_PKG_FILENAME);
+		STR_CASE(CELL_CROSS_CONTROLLER_ERROR_INVALID_SIG_FILENAME);
+		STR_CASE(CELL_CROSS_CONTROLLER_ERROR_INVALID_ICON_FILENAME);
+		STR_CASE(CELL_CROSS_CONTROLLER_ERROR_INVALID_VALUE);
+		STR_CASE(CELL_CROSS_CONTROLLER_ERROR_PKG_FILE_OPEN);
+		STR_CASE(CELL_CROSS_CONTROLLER_ERROR_SIG_FILE_OPEN);
+		STR_CASE(CELL_CROSS_CONTROLLER_ERROR_ICON_FILE_OPEN);
+		STR_CASE(CELL_CROSS_CONTROLLER_ERROR_INVALID_STATE);
+		STR_CASE(CELL_CROSS_CONTROLLER_ERROR_INVALID_PKG_FILE);
+		STR_CASE(CELL_CROSS_CONTROLLER_ERROR_INTERNAL);
+		}
 
-struct CellCrossControllerParam
-{
-	vm::bcptr<char> pPackageFileName;
-	vm::bcptr<char> pSignatureFileName;
-	vm::bcptr<char> pIconFileName;
-	vm::bptr<void> option;
-};
+		return unknown;
+	});
+}
 
-struct CellCrossControllerPackageInfo
+error_code cellCrossControllerInitialize(vm::cptr<CellCrossControllerParam> pParam, vm::cptr<CellCrossControllerPackageInfo> pPkgInfo, vm::ptr<CellCrossControllerCallback> cb, vm::ptr<void> userdata) // LittleBigPlanet 2 and 3
 {
-	vm::bcptr<char> pTitle;
-	vm::bcptr<char> pTitleId;
-	vm::bcptr<char> pAppVer;
-};
+	cellCrossController.todo("cellCrossControllerInitialize(pParam=*0x%x, pPkgInfo=*0x%x, cb=*0x%x, userdata=*0x%x)", pParam, pPkgInfo, cb, userdata);
 
-using CellCrossControllerCallback = void(s32 status, s32 errorCode, vm::ptr<void> option, vm::ptr<void> userdata);
+	// TODO
+	//if (something)
+	//{
+	//	return CELL_CROSS_CONTROLLER_ERROR_INVALID_STATE;
+	//}
 
-s32 cellCrossControllerInitialize(vm::cptr<CellCrossControllerParam> pParam, vm::cptr<CellCrossControllerPackageInfo> pPkgInfo, vm::ptr<CellCrossControllerCallback> cb, vm::ptr<void> userdata) // LittleBigPlanet 2 and 3
-{
-	cellCrossController.todo("cellCrossControllerInitialize(pParam=*0x%x, pPkgInfo=*0x%x, cb=*0x%x, userdata=*0x%x)", pParam, pPkgInfo, cb, userdata);
+	if (!pParam || !pPkgInfo)
+	{
+		return CELL_CROSS_CONTROLLER_ERROR_INVALID_VALUE;
+	}
+
+	if (!memchr(pParam->pPackageFileName.get_ptr(), '\0', CELL_CROSS_CONTROLLER_PARAM_FILE_NAME_LEN))
```

Comment:
I had check if the max size is with or without null terminator (i.e. without means `memchr(pParam->pPackageFileName.get_ptr(), '\0', CELL_CROSS_CONTROLLER_PARAM_FILE_NAME_LEN + 1)`.
Also check if pPackageFileName is nullptr, to avoid regression on dereferencing nullptr.
Same for other members.